deleteAndLoadSnapsets.sh

```diff
@@ -0,0 +1,79 @@
+#!/bin/sh
+
+STATIC_DATA_DIR=${STATIC_DATA_DIR:-/home/node/universal/testData/dbData}
+BUILD_DATA_DIR=${BUILD_DATA_DIR:-/home/node/universal/build/dbData/snapset}
+
+log() {
+  echo "$(date +'%Y-%m-%d %H:%M:%S') - $1"
+}
+
+warm_indices(){
+  log "Warming indices..."
+
+  for view in $(curl -s $COUCHDB_URL/_design/views/ | jq -r '.views | keys[]'); do
+    curl -fsS $COUCHDB_URL/_design/views/_view/$view >/dev/null
+  done
+
+  log "Finished warming indices..."
+}
+
+# Verify variables
+if [ -z "$COUCHDB_URL" ]; then
+  echo "COUCHDB_URL environment variable must be defined"
+  exit 1
+fi
+
+if [ ! -d "$STATIC_DATA_DIR" -o ! "$(ls -A $STATIC_DATA_DIR/*.json)" ]; then
+  echo "STATIC_DATA_DIR ($STATIC_DATA_DIR) does not exist or does not contain data, using universal's 'testData/dbData' as the default"
+  STATIC_DATA_DIR=./testData/dbData
+fi
+
+if [ ! -d "$BUILD_DATA_DIR" -o ! "$(ls -A $BUILD_DATA_DIR/*.json)" ]; then
+  echo "BUILD_DATA_DIR ($BUILD_DATA_DIR) does not exist or does not contain data, using universal's 'build/dbData/snapset' as the default"
+  BUILD_DATA_DIR=./build/dbData/snapset
+fi
+
+COUCHDB_URL_SANITIZED=`echo "$COUCHDB_URL" | sed -e 's,\(://\)[^/]*\(@\),\1<SENSITIVE>\2,g'`
+
+log "Starting"
+log "CouchDB: $COUCHDB_URL_SANITIZED"
+log "Clear index: $CLEAR_INDEX"
+log "Static: $STATIC_DATA_DIR"
+log "Build: $BUILD_DATA_DIR"
+log "Working directory: `pwd`"
+
+# Set up universal
+git clone --depth 1 --branch GPII-3138 https://github.com/klown/universal.git
```

Comment:
@stepanstipl One of the requirements of the new data loader is that it run whenever the snapset preferences inputs change or the '_design/views' input changes.  That happens when either of those changes are committed to universal (master).  If universal was cloned only when the dataloader docker image is built, then it has a fixed version of the preferences and views, and it won't always update CouchDB with the latest.  That was the rationale for moving the clone operation inside the image so it happens every time the dataloader runs.  That guarantees that the latest snapsets and views will be used.

However, that may not be the only way.  One way would be to move cloning to the docker image build step.  If the dataloader's docker image was re-built (and rerun) every time a snapset/views change is committed, that would be okay.  But that isn't much of an improvement.  It suffers the same slowness and github and npm issues you mentioned.

Another thought:  Is the universal Docker image re-built every time there is a commit to universal?  If so, that may be when the dataloader updates CouchDB.  Also, the code here could be moved into universal's codebase.

The earlier discussion about dupicating universal was not resolved because we were trying to get the new dataloader ready for the August demo.  So, we punted the duplication issue for the time being.  Looks like it's time to reconsider.

Thanks for pointing out that the clone points to the klown repository.  I said something along those lines at last week's architecture meeting, but it's good to repeat that point.  There are analogous issues in my universal pull, where [`vagrantCloudBasedContainers.sh`](https://github.com/GPII/universal/blob/2c48ca5217442ea94263e1d9148a419db008b21b/scripts/vagrantCloudBasedContainers.sh#L45).sh references my docker repository, not gpii's.

Reply:
@stepanstipl I have no objection to moving everything into universal, but I would like @cindyli's and @mrtyler's input.  Also, I'm wrestling with some questions/issues:

- what code is moved into universal?  Is it just the `deleteAndLoadSnapsets.sh` script?  What happens to the `Dockerfile` and docker container?  Do they just vanish?  Is its functionality replaced by running something inside the universal image?
- the `vagrantCloudBasedContainers.sh` script in universal [will change](https://github.com/GPII/universal/blob/d9ac25e3d3e9a763406d3fae375fe616b1079c23/scripts/vagrantCloudBasedContainers.sh#L85) if the dataloader docker image is no more.  It's not clear what takes its place.  Maybe just execute `deleteAndLoadSnapsets.sh` from within `vagrantCloudBasedContainers.sh`?
- If the dataloader docker container vanishes, does [pull 83](https://github.com/gpii-ops/gpii-infra/pull/83) vanish as well?  Its purpose it to set `CLEAR_INDEX` so the dataloader does not flush the database when it runs.  If pull 83 is no longer needed, what takes its place -- what sets `CLEAR_INDEX`?
- in the kubernetes cluster, the current dataloader docker image runs to completion and then terminates.  If the status code is zero on termination, it stays terminated.  However, if its status code is non-zero, it restarts and runs again.  This repeats until the dataloader succeeds.  This is different compared to the flowmanager which is deployed and keeps on going.  A reason why the dataloader fails is a race condition where CouchDB is also starting up and isn't ready for any request.  The dataloader fails, but starts again, and eventually succeeds when CouchDB is ready.

I'm thinking about these issues and if I have any insights, I'll write more here.  Thanks.

Reply:
thanks @klown, to answer your questions:

> what code is moved into universal? Is it just the deleteAndLoadSnapsets.sh script? What happens to the Dockerfile and docker container? Do they just vanish? Is its functionality replaced by running something inside the universal image?

Just `deleteAndLoadSnapsets.sh` and whatever portions from README are relevant. The `Dockerfile` is not needed as there's already one for universal, that seems to contain all the needed dependencies.

> the vagrantCloudBasedContainers.sh script in universal will change if the dataloader docker image is no more. It's not clear what takes its place. Maybe just execute deleteAndLoadSnapsets.sh from within vagrantCloudBasedContainers.sh?

Yes, that sounds like the correct solution.

> If the dataloader docker container vanishes, does pull 83 vanish as well? Its purpose it to set CLEAR_INDEX so the dataloader does not flush the database when it runs. If pull 83 is no longer needed, what takes its place -- what sets CLEAR_INDEX?

https://github.com/gpii-ops/gpii-infra/pull/83/files will not vanish, we will still set the env variable, just the image used in the infra part (`shared/charts/gpii-dataloader/templates/job.yaml`) will be the universal one and we will call the `deleteAndLoadSnapsets.sh` as a command. I'll update that PR once we agree on the move.

> in the kubernetes cluster, the current dataloader docker image runs to completion and then terminates. If the status code is zero on termination, it stays terminated. However, if its status code is non-zero, it restarts and runs again. This repeats until the dataloader succeeds. This is different compared to, say, the flowmanager which is deployed and keeps on going. A reason why the dataloader fails is a race condition where CouchDB is also starting up and isn't ready for any request. The dataloader fails, but starts again, and eventually succeeds when CouchDB is ready.

This behaviour will stay. We will still have the dataloader batch Job, just this job will use the universal image.